### PR TITLE
[mlir][linalg] Add more precise memory effects to linalg op

### PR DESCRIPTION
**Pull request by cxy-1993**
This patch add more precise memory effect to linalg op. Including the following points:
1. Remove the read side effects for operands that are not used.
2. Set the effect for all side effects to "full".

### CODE REVIEW

File: mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp [hunk 1]
```diff
@@ -1103,28 +1103,39 @@ ParseResult GenericOp::parse(OpAsmParser &parser, OperationState &result) {
 static void getGenericEffectsImpl(
     SmallVectorImpl<SideEffects::EffectInstance<MemoryEffects::Effect>>
         &effects,
-    ValueRange results, const ValueRange inputOperands,
+    LinalgOp linalgOp, ValueRange results, const ValueRange inputOperands,
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:1106]: Can you remove `inputeOperands` and `outputOperands` from the signature? They can be queried from `linalgOp`.

Comment (cxy-1993) [reply mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp]: Thanks for your advice. I have remove duplicated parameters from signature.

File: mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp [hunk 2]
```diff
@@ -1122,29 +1122,38 @@ ParseResult GenericOp::parse(OpAsmParser &parser, OperationState &result) {
...
+  unsigned usedOutputSize =
+      linalgOp.getOpOperandsMatchingBBargs().size() - inputOperandSize;
+
+  for (auto [index, operand] : llvm::enumerate(linalgOp.getDpsInits())) {
     if (!llvm::isa<MemRefType>(operand.getType()))
       continue;
-    effects.emplace_back(MemoryEffects::Read::get(), operand,
-                         SideEffects::DefaultResource::get());
-    effects.emplace_back(MemoryEffects::Write::get(), operand,
+    if (index < usedOutputSize &&
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:1142]: Why is this needed? `payloadUsesValueFromOperand` should be sufficient. Are you seeing a crash/out-of-bounds access in `getMatchingBlockArgument` without this change? In that case we should update `getMatchingBlockArgument` such that it returns a "null" BlockArgument. And `payloadUsesValueFromOperand` should return `false` is such a case.

Comment (cxy-1993) [reply mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp]: You are right, thank you for the reminder. I have modified the code according to your comment.

File: mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp [hunk 3]
```diff
@@ -2494,8 +2496,20 @@ SoftmaxOp::reifyResultShapes(OpBuilder &b,
 void SoftmaxOp::getEffects(
     SmallVectorImpl<SideEffects::EffectInstance<MemoryEffects::Effect>>
         &effects) {
-  getGenericEffectsImpl(effects, getOperation()->getResults(), getDpsInputs(),
-                        getDpsInits());
+  for (auto operand : getDpsInputs()) {
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:2499]: Why do we need to special case this instead of using `getGenericEffectsImpl`?

Comment (cxy-1993) [reply mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:2499]: Because only linalg.softmax does not have the LinalgOp Interface...

**Review by matthias-springer — APPROVED** [review]

**Comment by cxy-1993** [comment]
Counld you please help me merge this patch into master? @matthias-springer

File: mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td [hunk 4]
```diff
@@ -289,6 +289,9 @@ def MapOp : LinalgStructuredBase_Op<"map", [
 
     bool payloadUsesValueFromOperand(OpOperand * opOperand) {
       if (isDpsInit(opOperand)) return false;
+      if (getOperation()->getRegion(0).empty()) {
+        return true;
+      }
```
Comment (ftynse) [inline mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:294]: I don't quite get this. When the region is empty (is this even allowed?), the operand is considered used? But when it has a terminator not reading from the corresponding block argument, the operand is not considered used? Why? If we assume it is read anyway, it should be marked as such in both cases.

File: mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp [hunk 5]
```diff
@@ -1103,29 +1103,35 @@ ParseResult GenericOp::parse(OpAsmParser &parser, OperationState &result) {
...
-  for (auto operand : inputOperands) {
+    LinalgOp linalgOp) {
+  SmallVector<Value> inputOperands = linalgOp.getDpsInputs();
+  for (auto [index, operand] : llvm::enumerate(inputOperands)) {
     if (!llvm::isa<MemRefType>(operand.getType()))
       continue;
-    effects.emplace_back(MemoryEffects::Read::get(), operand,
-                         SideEffects::DefaultResource::get());
+    if (linalgOp.payloadUsesValueFromOperand(&linalgOp->getOpOperand(index))) {
+      effects.emplace_back(MemoryEffects::Read::get(), operand, 0, true,
```
Comment (ftynse) [inline mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:1112]: Nit: please use this style `/*argumentName=*/value` so `0` and `true` don't appear as magic constants here.

File: mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp [hunk 6]
```diff
@@ -2494,8 +2496,20 @@ SoftmaxOp::reifyResultShapes(OpBuilder &b,
...
         &effects) {
-  getGenericEffectsImpl(effects, getOperation()->getResults(), getDpsInputs(),
-                        getDpsInits());
+  for (auto operand : getDpsInputs()) {
+    if (!llvm::isa<MemRefType>(operand.getType()))
+      continue;
+    effects.emplace_back(MemoryEffects::Read::get(), operand,
+                         SideEffects::DefaultResource::get());
+  }
+  for (auto operand : getDpsInits()) {
```
Comment (ftynse) [inline mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:2505]: Nit: please expand auto unless the type is obvious from the line context.

**Review by ftynse — CHANGES_REQUESTED** [review]

File: mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td (hunk 4 above)
Comment (cxy-1993) [reply mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:294]: > I don't quite get this. When the region is empty (is this even allowed?), the operand is considered used? But when it has a terminator not reading from the corresponding block argument, the operand is not considered used? Why? If we assume it is read anyway, it should be marked as such in both cases.

Thank you for your suggestion! For this issue, here is my understanding: When region is empty, it usually occurs in scenarios where region does not need to interpret the semantics of op, such as linalg.map:

```
     %mapped = linalg.map { arith.addf }                                              
              ins(%arg0, %arg1 : tensor<10x100xf32>, tensor<10x100xf32>)           
              outs(%map_init : tensor<10x100xf32>)
```

In the example above, the linalg.map has an empty region. This is because, with arith.add present, we do not need to provide additional interpretation for this op, as its implementation simply involves adding the two input operands. In such scenarios where no interpretation is required, we consider all operands to be used in accordance with the semantics of the op.

In linalg operations implemented with regions, it is evident that when the block arguments corresponding to operands are not used, it indicates that they are not being read.

Comment (matthias-springer) [reply mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:294]: `linalg.map { arith.addf }` is just a pretty-printed version. It has a region with an `arith.addf` and `linalg.yield` terminator.

Comment (cxy-1993) [reply mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:294]: > `linalg.map { arith.addf }` is just a pretty-printed version. It has a region with an `arith.addf` and `linalg.yield` terminator.

You are right. I reviewed why I thought this way: before this, I didn't add handling for when region was empty, leading to some core dump issues during part of the bufferization tests. This happened because the 'cloneWithoutRegions' function was called during the bufferization process, and in this function, the side effects of 'op' were queried, unfortunately, at this time region was still empty. So, I'd like to solicit opinions from both of you: in this situation, can we conservatively assume that all operands can be read and written to?  @matthias-springer @ftynse

Comment (cxy-1993) [reply mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:294]: Hi @matthias-springer @ftynse,I wanted to send a friendly reminder to re-review the proposal above at your earliest convenience and hope to hear your thoughts on this issue.

Comment (matthias-springer) [reply mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:294]: Why does this lead to a crash? The bufferization impl for LinalgOps calls `cloneWithoutRegions`, but inlines the original region immediately afterwards. There should be no call to `payloadUsesValueFromOperand` in-between. (Even if there is one, this function would just return `false` and should not crash.)

Basically, the op is invalid without a region body. Generally speaking, helper functions on an op are allowed to function incorrectly if the op is invalid.

Comment (cxy-1993) [reply mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:294]: > Why does this lead to a crash?

When the cloneWithoutRegions function is called, the bufferize rewriter will invoke the notifyOperationInserted function. This function will query the side effects of the operation to count the number of memory allocations. Because this patch will call payloadUsesValueFromOperand when querying side effects, and the payloadUsesValueFromOperand of the map operation is implemented as follows:

```
bool payloadUsesValueFromOperand(OpOperand * opOperand) {                                                                                                                                  
   if (isDpsInit(opOperand)) return false;                                                                                                                                                                                                                                                                                                                     
  return !getMatchingBlockArgument(opOperand).use_empty();
}
```
This will fetch the block inside the op's region, causing a core dump.  

> Basically, the op is invalid without a region body. Generally speaking, helper functions on an op are allowed to function incorrectly if the op is invalid.

So should we prohibit querying side effects before the block is inserted?

Comment (matthias-springer) [reply mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:294]: Oh, I see. The op is already inserted even though it was not fully constructed yet (region missing). Ideally, we would change the `BufferizableOpInterface` impl., so that we insert the op only when it is fully constructed.

I think it's important to keep the `LinalgOp` interface clean and we should have to account for edge cases in the bufferization there.

Could you change the implementation in `Linalg/Transforms/BufferizableOpInterfaceImpl.cpp` such that it clones the op without regions without a builder. Basically using `Operation::create(OpState)` (instead of `cloneOpWithoutRegions`). Then inline the region. Then `rewriter.insert`. Then the callback in the bufferization will be triggered on a fully constructed and valid op.

Comment (cxy-1993) [reply mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td]: I agree with your perspective, we should keep the implementation within the interface clean. I have already modified the code according to your comment, thanks for your feedback.

**Review by matthias-springer — APPROVED** [review]

File: mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td (hunk 4 above)
Comment (cxy-1993) [reply mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td]: Hi @ftynse, do you have any further comments on the modifications proposed in this patch?

**Comment by cxy-1993** [comment]
Our friend ftynse hasn't had time to work on this patch recently, and the issues he raised seem to have been resolved. Could you please help me merge this PR?  @matthias-springer
